### PR TITLE
deps.windows: Update dependencies for CMake 3.0 upgrade

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description

* Disables x86 builds for all dependencies not required by x86-only modules.
* Moves luajit and Swig files into their canonical locations expected by default CMake find rules.
* Uses VSSetup PowerShell module to find Visual Studio installations in a more portable manner (also supports ARM64 Windows hosts)
* Creates obs-deps `VERSION` file in Windows packages

### Motivation and Context
* Enables compatibility with CMake 3.0 upgrade for Windows
* Removes complexity for CMake finders in said upgrade (canonical paths are used)
* Enables use of a single x64 project for Windows

### How Has This Been Tested?
Tested builds on Windows 11

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [inline utils.pwsh/Setup-Target.ps1:86]: If a user has VS2022 Stable and Preview both installed, which will it select? The system I would normally check this on is currently not up. The docs say, "By default, only releases are shown." unless you specify the `-Prerelease` flag, but a confirmation would be nice. If this does indeed only list stable releases, then this would be a slight behavior change, as we previously selected the first available from 2022 Release, 2022 Preview, 2019 Release. This should be fine on CI where only Release is available, though it could break on systems that only have VS2022 Preview installed.

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [inline utils.pwsh/Setup-Target.ps1:83]: Are there any permissions required to install this module in this scope? Are there any known reasons it could fail to install? Should we handle a failure case, or will the script bail out on its own?

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [inline utils.pwsh/Setup-Target.ps1:42]: Does this actually need the type specified, or is it just good practice? Checking both of these gives me the same output:
```
$VisualStudioData.InstallationVersion.GetType()
([System.Version] $VisualStudioData.InstallationVersion).GetType()
```

**Review by RytoEX — CHANGES_REQUESTED** [review]

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:86]: By default it should only pick up normal releases sorted by version and installation date. Prerelease versions are only included when the `-prerelease` or `-all` switch is added. This mirrors the behaviour of `vswhere` (which is on purpose iirc).

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:83]: None listed by Microsoft - it might prompt for installation permission when run on a local system, but that'd be something I would leave to a user by design and not override.

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:42]: It's good practise, because the cast makes it a version even if it's just a string - without it we'd have to check for the type before accessing the `Member` variable to avoid an exception.

Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:42]: I mean that it doesn't seem to materially change the type or output of this particular data returned from VSSetup, which is already a `Version`. By comparison, the type for `InstallationLocation` from the old method _was_ a string, so it was necessary there. I'm not super fussed about it, it just seemed unnecessary with the new method.
```
PS C:\> $VisualStudioData.InstallationVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
17     5      33530  505

PS C:\> $VisualStudioData.InstallationVersion.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Version                                  System.Object

> > ([System.Version] $VisualStudioData.InstallationVersion)

Major  Minor  Build  Revision
-----  -----  -----  --------
17     5      33530  505

PS C:\> ([System.Version] $VisualStudioData.InstallationVersion).GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Version                                  System.Object

PS C:\> $VisualStudioData = Get-CimInstance MSFT_VSInstance
PS C:\> $VisualStudioData[0].InstallLocation
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community

PS C:\> $VisualStudioData[0].InstallLocation.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     String                                   System.Object
```

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:83]: Sounds good to me.

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:86]: Hmm. Our old behavior checked for VS2022 Release, then VS2022 Preview, then VS2019 Release. If we want to mimic that behavior, we might have to call this multiple times for each specific version. Otherwise, we abandon that behavior and accept that it will fail on systems that only have VS2022 Preview installed.

Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:86]: I think if we add `-prerelease` the choice would still be based on version number. So If VS2022 17.1 and VS2022 17.1 Preview are installed (not that this is a probable situation), I guess the release version is "more recent". VS2022 17.2 Preview would _be_ more recent however. So it would always chose whatever has the higher version number.

To that end we could just add that one to also include preview versions.

Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:86]: If 17.6 Release and 17.7 Preview were both installed and `-Prerelease -Latest` was specified, I imagine it would pick 17.7 Preview. The real question is what if the user has 17.7 Release and 17.7 Preview installed.

That said, the previous selection process didn't even bother taking this into account, it simply preferred:
1. VS2022 Release (any version)
2. VS2022 Preview (any version)
3. VS2019 Release (any version)

If you had VS2022 17.3 Release and VS2022 17.7 Preview installed, it would select VS23022 17.3 Release. This was, presumably, to allow devs to have the Release version that we targeted installed while also having some other Preview version installed if they needed it for other dev purposes outside of OBS.

Will think on it a bit.

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:83]: As an aside, this will probably cause a user prompt, which is fine, but something to be aware of.
```powershell
PS C:\> Install-Module VSSetup -Scope CurrentUser

Untrusted repository
You are installing the modules from an untrusted repository. If you trust this repository, change its
InstallationPolicy value by running the Set-PSRepository cmdlet. Are you sure you want to install the modules from
'PSGallery'?
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "N"):
```

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:86]: Release and Preview would probably differ in their patch or build versions, so that would be taken into account. So if Release is build 6521 and Preview is 6532, then Preview "wins".

I would expect this to be a very simple version number comparison, only one variant includes the version numbers of preview builds and the other doesn't.

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:83]: Correct. Personally I feel it's a good choice not to suppress this, but let users handle this on their local machines. It's in any case simpler than whatever script switches we could provide.

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:86]: Tested with 17.5 Release and 17.7 Preview.
```pwsh
PS C:\> Get-VSSetupInstance | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest

InstanceId          : 142e8861
DisplayName         : Visual Studio Community 2022
InstallationVersion : 17.5.33627.172
InstallationPath    : C:\Program Files\Microsoft Visual Studio\2022\Community
InstallDate         : 11/1/2022 1:38:25 PM


PS C:\> Get-VSSetupInstance -Prerelease | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest

InstanceId          : 3d2ad485
DisplayName         : Visual Studio Community 2022
InstallationVersion : 17.7.33711.374
InstallationPath    : C:\Program Files\Microsoft Visual Studio\2022\Preview
InstallDate         : 5/20/2023 1:40:40 PM
```

Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:86]: That at least worked as expected. It is still just a little unfortunate that it differs from current behavior. The only way I can think of to get the current behavior is to do multiple calls.

```pwsh
$VSInstance = Get-VSSetupInstance | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest
// some null/empty/false check
$VSInstance = Get-VSSetupInstance -Prerelease | Select-VSSetupInstance -Version '[16.0,18.0)' -Latest
```

Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:86]: I'd argue that the old behaviour was actually broken insofar as even a more recent VS 2022 preview version would not have been picked up.

The only scenario that code was made for was for the transitional period for when there was only a preview version of VS 2022 but a release was about to happen and we wanted the jobs not to suddenly break on CI.

Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:86]: > The only scenario that code was made for was for the transitional period for when there was only a preview version of VS 2022 but a release was about to happen and we wanted the jobs not to suddenly break on CI.

That's not what it was for. GitHub Actions images never have VS Preview builds, only Release builds. If I recall correctly, it was intended to provide build stability for _local_ builds where a dev might have multiple versions of VS installed for _some_ reason (work, other dev projects, etc.), but still wanted to be able to build obs-deps.

Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:86]: Fixed via adding explicit prerelease switch to main build script toggling inclusion of Preview builds in version checks.

File: deps.windows/50-swig.ps1
Comment (RytoEX) [inline deps.windows/50-swig.ps1:108]: Is there really nothing left in `$($ConfigData.OutputPath)/swig` to keep?

Comment (PatTheMav) [reply deps.windows/50-swig.ps1:108]: Not that I'm aware of - anything swig needs is its binary and the language-specific modules which sit in the `share` subdirectory.

Afaik Swig also just looks for a `Lib` directory adjacent to the binary exclusively on Windows, because there is no POSIX directory standard - more recent versions might _also_ support that, but it was the search location with the highest priority.

**Review by RytoEX — APPROVED** [review]
Seems generally okay to me at a glance. Will test locally once my dev environment is working again. May merge a couple of other PRs first, but will try to keep it minimal.

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [inline utils.pwsh/Setup-Target.ps1:43]: This will need adjusted to not overwrite the change to use regex.

**Review by RytoEX — CHANGES_REQUESTED** [review]

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:43]: That change is intended - the old variant was based on the output of the CMI provider, the update uses `VSSetup` and its output instead.

@Absolucy would you be able to check what the `DisplayName` given by `VSSetup` is in your setup?

Comment (RytoEX) [reply utils.pwsh/Setup-Target.ps1:43]: There's no need. I checked this already before merging #178.
```
PS C:\> Get-VSSetupInstance -Prerelease:$false

InstanceId          : 142e8861
DisplayName         : Visual Studio Community 2022
InstallationVersion : 17.6.33723.286
InstallationPath    : C:\Program Files\Microsoft Visual Studio\2022\Community
InstallDate         : 11/1/2022 1:38:25 PM

InstanceId          : a63227f3
DisplayName         : Visual Studio Build Tools 2022
InstallationVersion : 17.6.33723.286
InstallationPath    : C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools
InstallDate         : 5/30/2023 1:17:32 PM
```

File: deps.windows/60-python.ps1
Comment (RytoEX) [inline deps.windows/60-python.ps1:25]: Why the capitalization change here?

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [inline utils.pwsh/Setup-Target.ps1:80]: `$Env` vs. `$env`?

File: utils.pwsh/Setup-Target.ps1
Comment (RytoEX) [inline utils.pwsh/Setup-Target.ps1:43]: There should either be a space at the end of this parenthetical, or we should remove the space at the beginning. We're not 100% consistent with this throughout, but we don't need to do a consistency sweep in this PR.

File: deps.windows/60-python.ps1
Comment (RytoEX) [inline deps.windows/60-python.ps1:36]: What's the reason for this change?

File: Build-Dependencies.ps1
Comment (RytoEX) [inline Build-Dependencies.ps1:74]: What is the intent here?

As an aside, although this `if` is not new, it seems like the `if` test will always return `! ( $true )`, since `$Targets` is set to `@($script:Target)` just above. Perhaps I've misunderstood something.

File: .github/workflows/main.yaml
Comment (RytoEX) [inline .github/workflows/main.yaml:437]: What's the purpose of removing this?

**Review by RytoEX — CHANGES_REQUESTED** [review]
A quick visual look over the files before I dig back into testing.

File: deps.windows/60-python.ps1
Comment (PatTheMav) [reply deps.windows/60-python.ps1:25]: Mainly that `env` (like `script` or other scopes) are not capitalised in Microsoft's documentation, and it also makes it aesthetically easier to read (as the _actually_ capitalised variable name stands out better).

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:80]: See above.

File: utils.pwsh/Setup-Target.ps1
Comment (PatTheMav) [reply utils.pwsh/Setup-Target.ps1:43]: I think we use spaces in conditionals, but don't use them when enclosing expressions.

File: deps.windows/60-python.ps1
Comment (PatTheMav) [reply deps.windows/60-python.ps1:36]: See above.

File: Build-Dependencies.ps1
Comment (PatTheMav) [reply Build-Dependencies.ps1:74]: `continue` doesn't work as it would in a `foreach` blog - instead it continues the execution of the current script block and breaks the entire script.

`ForEach-Object` is literally borrowing behaviour from functional languages, in which the script block is an anonymous script block. Skipping an element thus requires simply returning from the current block early.

`$Stages` exists outside of that scope and is stateful, so an early return requires cleaning up all stage functions created by the dot import of the script file. Without this, the `Patch` function of an earlier dependency would be executed over and over again until a later dependency overwrites it with its own `Patch` step.

$Targets is overwritten with every dot import of a new dependency script and needs to be evaluated on every run of the anonymous script block.

File: .github/workflows/main.yaml
Comment (PatTheMav) [reply .github/workflows/main.yaml:437]: The NTV2 SDK cannot be built as a shared library. Thus checking for the shared workflow variable is superfluous.

File: deps.windows/60-python.ps1
Comment (RytoEX) [reply deps.windows/60-python.ps1:25]: Discussed off-thread. Microsoft's documentation itself is inconsistent with capitalization.

While I'd prefer that style/consistency changes be done in a single sweep in their own PR, I don't want to hold this up with trying to excise and separate these changes.

File: Build-Dependencies.ps1
Comment (RytoEX) [reply Build-Dependencies.ps1:74]: Discussed further off-thread.

**Review by RytoEX — APPROVED** [review]
Tested against obs-studio git and the OBS Studio Windows CMake 3 PR. Successfully built and ran OBS in both cases with my test scene collection, so this should be good to go.

**Pull request merged**
